### DISCUSSION TITLE
Help me understand how to use this plugin

### DISCUSSION THREAD

**Discussion by charlie39** [kylechui/nvim-surround#51] [Q&A] [answered]
``` -- Define pairs based on function evaluations!
            ["i"] = function()
                return {
                    require("nvim-surround.utils").get_input(
                        "Enter the left delimiter: "
                    ),
                    require("nvim-surround.utils").get_input(
                        "Enter the right delimiter: "
                    )
                }
            end,
            ["f"] = function()
                return {
                    require("nvim-surround.utils").get_input(
                        "Enter the function name: "
                    ) .. "(",
                    ")"
                }
            end,
        },
```
I kind of know what these are but i am not sure how i would use them?

```
separators = {
            ["'"] = { "'", "'" },
            ['"'] = { '"', '"' },
            ["`"] = { "`", "`" },
        },
``` 
like separators in a list? shouldn't comma be a more appropriate separator? and how would i use them?

It would be nice if it support count as well , like ```ysa2w"``` -> qoute  two words

**Comment by kylechui** [comment, marked as answer]
Hi there @charlie39! I guess I never really gave too much information on how to use my plugin because I assumed my users were primarily converting from [`vim-surround`](https://github.com/tpope/vim-surround) or [`vim-sandwich`](https://github.com/machakann/vim-sandwich), so I'll explain a bit here (also see `:h nvim-surround`). The main idea is that for every character that you have in your `pairs` table, you should have a table of two "delimiters" that will surround your selections, one for the left and one for the right.

The idea behind associating characters with *functions* instead of a normal table, is that you can call into the Lua API  for more complex, dynamic surrounds, as opposed to always surrounding with the exact same thing every single time. In the two examples configured by default, they ask for user input for what should be put on the left or right of your selection.

As for the term "separators", I just wanted to make a distinction between delimiters that were *different* for left and right (pairs), and those that were the same for left and right (separators). Again, this is explained more in `:h nvim-surround`.

For the final question, the format for surrounding depends heavily on (Neo)vim's built-in text-object system, so the appropriate command would be `ys2aw"`, as `2aw` is a valid text-object, not `a2w`.

Hopefully this answers a few of your questions, and feel free to drop by and ask more later!

Edit: Checking your post's edit history, it might be worth it for you to check out `vim-surround`'s README for the basics on how to use this plugin.

Edit 2: The basic idea behind this plugin (and other similar surround plugins) follows Vim's grammar, e.g. `[verb][object]`. In this case, the provided "verbs" are `ys` (you surround/insert surround), `cs` (change surround), and `ds` (delete surround).
* For inserting surrounds, we use `ys[object][char]` to insert `char` around the given `object`, e.g. `ysiw)` surrounds the current word with parentheses.
* For deletions, we have `ds[char]` to delete the surrounding pair associated with `char`, e.g. `ds"` will delete the surrounding double quotes
* Finally for changes, we have `cs[to delete][to replace]`, to replace a given surrounding pair, e.g. `cs'{` will replace the surrounding single quotes with curly braces

This methodology extends to other, more complex mappings (at least for inserting), e.g. `ysi(i` will take the contents of the text inside the nearest set of parentheses, then query the user for what to surround the selected text with (on both left and right sides).

**Reply by charlie39** [reply]
Would be nice, if you cite similiar examples as you did on reddit here as well, so others like me may find it handy

**Reply by kylechui** [reply]
Do you mind checking out the second edit to see if that helps you digest the information a bit better?
